### PR TITLE
adapter: add end-to-end test for LaunchDarkly synchronization

### LINKED ISSUE #16583: Add a test for the LaunchDarkly system parameter sync
We need an end-to-end test for the code added as part of #15956.

The test should exercise both the initial sync in `Catalog::open` and the periodic sync loop (`system_parameter_sync`).

### PR DESCRIPTION
**Pull request by aalexandrov**
A follow-up of #16089 which makes some further changes required to add an end-to-end test and actually adds the end-to-end test.

### Motivation

  * This PR adds a known-desirable feature.

Closes #16583, #16587.

### Tips for reviewer

* ~I've disabled some large parts of `pipeline.template.yml` so I don't have to wait hours after every minor tweak. I will revert this before I merge.~ This is now present only in the nightlies config.
* I've taken the liberty to put a bunch of `type: ignore` hints on the `launchdarkly_api` imports to make `mypy` happy. I gave `stubgen -p launchdarkly_api` a try, but the result didn't work out of the box. If it is absolutely necessary to fix those lints the right way I will have to allocate some more time to do that, but I feel the right thing to do here is to ask the LD team to generate stubs for their Python client automated generation process.
* An open question for me is how exactly to handle the race conditions that occur between `ld_client.update_targeting` calls and the subsequent `c.testdrive` call. At the moment I have very sprinkled some `sleep(3)` statements, but for a more robust test we need to implement some retry logic. The options are: 
  * ~Make `ShowVariables` retry-able in `src/testdrive/src/action/sql.rs`.~
  * Do this ad-hoc in this specific test (given that `SHOW $var` will be deterministic in every other test I think this should be preferred for now). :heavy_check_mark: 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. See [the runs for this branch in nightlies section](https://buildkite.com/materialize/nightlies/builds?branch=aalexandrov%3Ald_sync_e2e_test).
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: ci/test/pipeline.template.yml
Comment (benesch) [inline ci/test/pipeline.template.yml:594]: I think we might want to make this be a nightly and release qualification test instead? Otherwise I worry we'll be hitting the LD API quite a lot, and any flakiness in the LD API will block merging.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:727]: Would you actually mind to just delete this entire `println!`? We're leaking secrets left and right here (the Cockroach passwords come in as env vars, for example). This made sense in the binary days, but these days, it's just as easy to pull this information from Kubernetes (if not easier).

**Review by benesch — COMMENTED** [review]
Are the LD feature flag names scarce resources? If so, should we add a job to ci/cleanup that periodically cleans them up?

File: ci/test/pipeline.template.yml
Comment (philip-stoev) [reply ci/test/pipeline.template.yml:594]: Yes please make it Nightly and then I will take care of any flakes going forward.

**Comment by philip-stoev** [comment]
> An open question for me is how exactly to handle the race conditions that occur between ld_client.update_targeting calls and the subsequent c.testdrive call. 

What race conditions are you worried about? Testdrive will retry your SHOW statement until it matches the expected value.

**Review by philip-stoev — APPROVED** [review]
Thank you! This test wins the award for the most elaborate test of 2022!

**Comment by aalexandrov** [comment]
> Are the LD feature flag names scarce resources?

According to [the pricing plan](https://docs.launchdarkly.com/home/billing/plans) the number of flags is unlimited. The billable resource are the users (identified by a LaunchDarkly user key), which is why all tests run under `anonymous-dev@materialize.com`.

> If so, should we add a job to ci/cleanup that periodically cleans them up?

The test is meant to be cleaning the feature flag it generates at the end of the test with this line:

```python
ld_client.delete_flag(LD_FEATURE_FLAG_KEY)
```

For the occasional flake we can also delete those feature flag manually (they are always tagged with `ci-test`). If this becomes too much of an overhead I'm happy to add a CI job.

File: test/launchdarkly/mzcompose.py
Comment (teskje) [inline test/launchdarkly/mzcompose.py:210]:
```suggestion
    A test-specific LaunchDarkly client that simulates a client modifying
```

**Review by teskje — APPROVED** [review]
LGTM, just found a couple typos in comments.

**Comment by benesch** [comment]
> I added this to a finally clause in the latest revision. If this does not work as expected and we start accumulating junk in LD I can add a CI job.

After enough test runs I guarantee that some small amount of junk will accumulate. Buildkite doesn't guarantee that the test will run to completion; the agent might get lost partway through, or a cancellation (e.g., due to a new push to the branch) might get forced.

It looks like the LD API exposes the `creationDate` of the feature flag, which will make it easy to write the cleanup job down the road. It's usually as simple as: list all feature flags, find any that are more than a day old and match the name "ci-test", and delete them.

**Comment by benesch** [comment]
(Btw, I didn't look at this PR in much detail, but don't wait for me for an approval! Deferring to @teskje.)

**Comment by aalexandrov** [comment]
> You should add the cleanup job to [`main`/ci/cleanup/pipeline.yml](https://github.com/MaterializeInc/materialize/blob/main/ci/cleanup/pipeline.yml?rgh-link-date=2022-12-12T13%3A24%3A03Z).

I added [an issue to track this](https://github.com/MaterializeInc/materialize/issues/16587).

**Pull request merged**
